src/coreclr/jit/assertionprop.cpp

```diff
@@ -3791,9 +3821,32 @@ AssertionIndex Compiler::optAssertionIsNonNullInternal(GenTree* op, ASSERT_VALAR
             return NO_ASSERTION_INDEX;
         }
 
-        ValueNum vn = op->gtVNPair.GetConservative();
+        // Look at both the top-level vn, and
+        // the vn we get by stripping off any constant adds.
+        //
+        ValueNum  vn     = vnStore->VNConservativeNormalValue(op->gtVNPair);
+        ValueNum  vnBase = vn;
+        VNFuncApp funcAttr;
 
-        // Check each assertion to find if we have a vn == or != null assertion.
+        while (vnStore->GetVNFunc(vnBase, &funcAttr) && (funcAttr.m_func == (VNFunc)GT_ADD))
```

Comment:
Ideally perhaps this while loop would be unnecessary, and VN creation would fold chains of constant adds into a single add.

Also note there is a similar loop on the gen side in `optCreateAssertion` which given a dereference through a byref, tries to walk the VN back to the inspiring gc ref.

Reply:
Also we should be able to get rid of the IR destructuring (see first part of `optAssertionProp_Ind` for example) for global prop as the VN destructuring should be equivalent, but when I did that I lost some of the diffs, and it creates more divergence between local prop and global prop.